[CSPM] Add support for XCCDF benchmarks

0intro opened the pull request:
### What does this PR do?

This change adds support for XCCDF benchmarks in the pkg/compliance package.

XCCDF is an XML format specifying security benchmarks, which is part of the [SCAP](https://csrc.nist.gov/projects/security-content-automation-protocol/) standards.

The XCCDF support relies on the [OpenSCAP](https://github.com/OpenSCAP/openscap) library.

### Motivation

The goal was to implement support of CIS benchmarks for Linux operating systems into the Datadog Agent.

Instead of writing the checks ourselves, we chose to rely on the automated checks available as part of the [ComplianceAsCode](https://github.com/ComplianceAsCode/content) project. The checks are written in [SCAP](https://csrc.nist.gov/projects/security-content-automation-protocol/) format and can be executed with the [OpenSCAP](https://github.com/OpenSCAP/openscap) library and tools.

### Additional Notes

### Possible Drawbacks / Trade-offs

The major drawback of including the OpenSCAP library is that it requires its own set of dependencies, adding 25 new software definitions to omnibus.

However, we tried to limit the size of the new dependencies, by removing the unnecessary files from the Agent installation, like documentation and static libraries. The total size of the newly installed shared libraries approximates 20 MB.

### Describe how to test/QA your changes

Build for RHEL:

```
cd datadog-agent
docker run -v "$PWD:/go/src/github.com/DataDog/datadog-agent" -v "/tmp/omnibus:/omnibus" -v "/tmp/opt/datadog-agent:/opt/datadog-agent" -v"/tmp/gems:/gems" --workdir=/go/src/github.com/DataDog/datadog-agent datadog/agent-buildimages:deb-x64 inv -e agent.omnibus-build --base-dir=/omnibus --gem-path=/gems
```

Or build for Ubuntu:

```
cd datadog-agent
docker run -v "$PWD:/go/src/github.com/DataDog/datadog-agent" -v "/tmp/omnibus:/omnibus" -v "/tmp/opt/datadog-agent:/opt/datadog-agent" -v"/tmp/gems:/gems" --workdir=/go/src/github.com/DataDog/datadog-agent datadog-agent-buildimages:rpm-x64 inv -e agent.omnibus-build --base-dir=/omnibus --gem-path=/gems
```

Run RHEL 7, 8 or 9:

```
docker run -v "/tmp/opt:/opt" -v "/:/host" -it registry.access.redhat.com/ubi7/ubi /bin/bash
```

Or run Ubuntu 20.04 or 22.04:

```
docker run -v "/tmp/opt:/opt" -v "/:/host" -it ubuntu:22.04 /bin/bash
```

Remove the `/.dockerenv` file, so OpenSCAP doesn't know it's running in a container:

```
rm -f /.dockerenv
```

Install curl (on RHEL):

```
yum install curl -y
```

Or install curl (on Ubuntu):

```
apt update && apt install curl -y
```

Install CSPM rules:

```
mkdir -p /etc/datadog-agent/compliance.d
curl -Ls https://github.com/DataDog/security-agent-policies/archive/refs/heads/jinroh/cspm-scap-rules.tar.gz | tar xz --strip-components=3 -C /etc/datadog-agent/compliance.d -f - security-agent-policies-jinroh-cspm-scap-rules/compliance
```

Configure datadog-agent:

```
cat << EOF >/etc/datadog-agent/datadog.yaml
api_key: <api_key>
site: datad0g.com
hostname: test-xccdf

security_agent:
  remote_tagger: false

compliance_config:
  enabled: true
  dir: /etc/datadog-agent/compliance.d
  xccdf:
      enabled: true
EOF
```

Run the compliance checks from the `security-agent`, on docker:

```
/opt/datadog-agent/embedded/bin/security-agent -c /etc/datadog-agent/datadog.yaml start
```

Run the compliance checks from the `security-agent`, on the host's operating system:

```
DOCKER_DD_AGENT=true /opt/datadog-agent/embedded/bin/security-agent -c /etc/datadog-agent/datadog.yaml start
```

Go to the [CSPM web page](https://dd.datad0g.com/security/compliance?query=framework%253Acis-ubuntu2204-level1&aggregation=rules&sort=ruleSeverity%252CfailedResources-desc) and verify that you can see the check results.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

lebauce committed:
[omnibus] Allow passing build tags to security-agent.build

lebauce committed:
[tasks/licenses] Escape quote in third party licenses CSV

lebauce committed:
[omnibus] Do not install some library .a archives

lebauce committed:
[CSPM] Allow setting filters on compliance rules

lebauce committed:
[CSPM] Allow setting per rule interval

jinroh committed:
[omnibus] Deactivate openscap build against arm7l machine

0intro committed:
[omnibus] Add OpenSCAP and its dependencies

OpenSCSAP is an implementation of the SCAP standards.
It provides tools and libraries to handle SCAP documents.

This change adds the following software definitions:

 - apt
 - attr
 - elfutils
 - file
 - glib
 - gmp
 - gnutls
 - libacl
 - libdb
 - libgcrypt
 - libgpg-error
 - liblz4
 - libselinux
 - libsepol
 - libtasn1
 - libxxhash
 - lua
 - m4
 - nettle
 - openscap
 - pcre2
 - popt
 - rpm
 - sqlite
 - util-linux
 - xmlsec
 - zstd

Some patches were required:

 - apt/cmake-bindir.patch: allow to update BIN_DIR CMake CACHE entry
 - apt/disable_arch_check.patch: disable architecture check
 - apt/disable_systemd.patch: remove dependency on udev and systemd
 - apt/no_doc.patch: don't install documentation
 - apt/triehash.patch: add the triehash tool, needed at build time
 - libselinux/fcntl_o_path.patch: fix build with old glibc
 - libselinux/ln_no_relative.patch: don't use relative symlink on installed libraries
 - libsepol/ln_no_relative.patch: don't use relative symlink on installed libraries
 - lua/nodoc.patch: don't install documentation
 - nettle/install_to_lib.patch: omit lib directory suffix
 - rpm/0001-Include-fcntl.patch: fix build

OpenSCAP includes a number of fixes, backported from maint-1.3
branch and also some fixes which are currently being reviewed.

We added -I. before -I#{install_dir}/embedded/include in CGO_CFLAGS,
so the Go zstd package use its own zstd.h instead of using the one
provided by the newly added zstd software.

This change is the result of a collaborative work between
Sylvain Baubeau, Pierre Guilleminot and David du Colombier.

Co-authored-by: Sylvain Baubeau <lebauce@gmail.com>
Co-authored-by: Pierre Guilleminot <pierre.guilleminot@datadoghq.com>

maycmlee commented on releasenotes/notes/cspm-xccdf-3d6cefbaaaa117b5.yaml:
```diff
@@ -0,0 +1,6 @@
+  - |
+    Add support for XCCDF benchmarks in CSPM.
+    A new configuration option, 'compliance_config.xccdf.enabled',
+    disabled by default, has been added, to enable XCCDF benchmarks.
```
```suggestion
    disabled by default, has been added for enabling XCCDF benchmarks.
```

0intro replied:
Fixed. Thanks!

maycmlee reviewed (commented):
Small suggestion

jinroh committed:
[CSPM] Add support for evaluating XCCDF rules in batch

jinroh committed:
[CSPM] Properly wrap libopenscap to lock all sessions in one OS thread

0intro committed:
[CSPM] Implement OpenSCAP session reuse

bkabrda commented on omnibus/config/software/apt.rb:
```diff
@@ -0,0 +1,96 @@
+  env["CXX"] = "/opt/gcc-#{ENV['GCC_VERSION']}/bin/g++"
+  env["CXXFLAGS"] += " -static-libstdc++ -std=c++11"
+  patch source: "no_doc.patch", env: env
+  patch source: "disable_arch_check.patch", env: env
```
Not a blocker for now, but in the foreseeable future, could you please add comments to all the patches on all software definitions that explain why we have them? For example, it is not clear to me why we disable the arch check for APT.

bkabrda commented on omnibus/config/software/apt.rb:
```diff
@@ -0,0 +1,96 @@
+  patch source: "disable_systemd.patch", env: env
+  patch source: "cmake-bindir.patch", env: env
+
+  if (!File.exist? '/usr/bin/triehash') && (!File.exist? '/usr/local/bin/triehash')
```
Is the `triehash` file something that we should add to our buildimages longer term? Again, this is not something I see as a blocker, but it looks like a build dependency that won't change (often) and could live there - it's not part of the distributed agent, correct?

0intro replied:
`triehash` is a build time dependency for building APT.

bkabrda replied:
Yup, so eventually it would be great to move it to buildimages, rather than have it in this repository.

0intro replied:
The question was raised at some point, but I finally preferred to install `triehash` as part of the apt build, since it's only used to build apt. My feeling was that the build images had a more general purpose.

lebauce replied:
I reopened the PR with only the triehash relevant bits : https://github.com/DataDog/datadog-agent-buildimages/pull/378

bkabrda replied:
Thanks for opening the PR! Because we're a bit short on time, I'll approve this PR with triehash file included and ask you to, later on, upgrade the buildimages and remove the file from here in a followup PR.

bkabrda commented on omnibus/config/software/datadog-agent.rb:
```diff
@@ -154,7 +156,8 @@
   # Security agent
   unless windows?
-    command "invoke -e security-agent.build --major-version #{major_version_arg}", :env => env
+    build_tags = linux? ? "" : ""
```
Is this an error? If I'm looking right, we always use `""` as `build_tags`?

0intro replied:
This will go away, since we won't link with the OpenSCAP library by default.

bkabrda reviewed (commented):
:wave: the PR mostly looks good, I left couple points inline to address/talk about. In addition, we need to add the `ship_source_offer true` line to all the new GPL/LGPL software.

0intro committed:
[CSPM] Remove xccdf_cgo.go

This change removes the implementation of XCCDF which relies
on the OpenSCAP library. Instead we use xccdf_no_cgo.go,
which relies on the OpenSCAP tool. This prevents the Agent
to link with OpenSCAP and its dependencies.

bkabrda commented on omnibus/config/software/datadog-agent.rb:
```diff
@@ -41,7 +43,7 @@
         "Python3_ROOT_DIR" => "#{install_dir}/embedded",
         "LDFLAGS" => "-Wl,-rpath,#{install_dir}/embedded/lib -L#{install_dir}/embedded/lib",
-        "CGO_CFLAGS" => "-I#{install_dir}/embedded/include",
+        "CGO_CFLAGS" => "-I. -I#{install_dir}/embedded/include",
```
Why do we need to add this include?

0intro replied:
This is a good question. I've already documented it in the commit message:

```
We added -I. before -I#{install_dir}/embedded/include in CGO_CFLAGS,
so the Go zstd package use its own zstd.h instead of using the one
provided by the newly added zstd software.
```

Before this change, we had to modify the name of the `zstd.h` file
in the Go [zstd](https://github.com/DataDog/zstd) package, so it doesn't conflicts with the newly installed 
`#{install_dir}/embedded/include/zstd.h`.

bkabrda replied:
Ah, thanks, I didn't notice the description in the commit message. Thanks!

0intro committed:
[CSPM] Add support for XCCDF benchmarks

This change adds support for XCCDF benchmarks in the pkg/compliance
package.

XCCDF is an XML format specifying security benchmarks, which is part
of the SCAP standards.

The XCCDF support relies on the OpenSCAP library, which is provided
as part of a previous change.

The XCCDF files should be located into the compliance directory,
specified by the "compliance_config.dir" configuration option.

Support of XCCDF benchmark is disabled by default, and can be
enabled by setting the new "compliance_config.xccdf.enabled"
configuration option.

This change is the result of a collaborative work between
Sylvain Baubeau, Pierre Guilleminot and David du Colombier.

Co-authored-by: Pierre Guilleminot <pierre.guilleminot@datadoghq.com>
Co-authored-by: David du Colombier <djc@datadoghq.com>

lebauce committed:
[CI] Temporarely allow a package size difference of 60MB

0intro commented:
The `macos_test` job is failing on:

```
 # github.com/DataDog/datadog-agent/pkg/network/encoding [github.com/DataDog/datadog-agent/pkg/network/encoding.test]
pkg/network/encoding/http.go:87:31: undefined: network.ConnectionKeysFromConnectionStats
pkg/network/encoding/http.go:102:28: undefined: network.ConnectionKeysFromConnectionStats
pkg/network/encoding/http2.go:45:28: undefined: network.ConnectionKeysFromConnectionStats
pkg/network/encoding/http2.go:101:31: undefined: network.ConnectionKeysFromConnectionStats
pkg/network/encoding/kafka.go:80:31: undefined: network.ConnectionKeysFromConnectionStats
pkg/network/encoding/kafka.go:94:28: undefined: network.ConnectionKeysFromConnectionStats
